design.md

```diff
@@ -0,0 +1,226 @@
+
+`-stop JOB-ID` Stops the job identified by `JOB-ID` and returns its exit code or an error if the provided job did not exist. It must be used to release the resources of the system.
+
+`-status JOB-ID` Returns the current state (Started or Done) of the job identified by `JOB-ID` or an error if the provided job did no exist.
+
+`-stdout JOB-ID` Writes the standard output of the given job to the standard output of this process, or returns an error if another instances of this process is already connected.
+
+`-stderr JOB-ID`	Writes the standard error of the given job to the standard output of this process, or returns an error if another instance of this process is already connected.
+
+### Example session
+
+```
+$ overseer-cli -server localhost:8888 -key my-key.key -cert server.crt -start echo Hello
+UNIQUE-JOB-ID
+
+$ overseer-cli -server localhost:8888 -key my-key.key -cert server.crt -status UNIQUE-JOB-ID
+Done
+
+$ overseer-cli -server localhost:8888 -key my-key.key -cert server.crt -stdout UNIQUE-JOB-ID
+Hello
+
+$ overseer-cli -server localhost:8888 -key my-key.key -cert server.crt -stop UNIQUE-JOB-ID
+0
+```
+
+
+## Resource control
+
+There are many ways to control the resources on a Linux system, alternatives considered were: `systemd-run`, `cpulimit`, `nice`, `ionice`,`cgroups`; but the `prlimit()` system call was chosen because it satisfies the requirements (being a system call and being able to enforce all the necessary limits). The proposed solution is to start the given job and then set the limits, the drawback being that the new process could exceed the resources before the limits are set, another drawback is that there is no reliable way to tell if a process exceeded the limits and got killed or not.
+
+The external package [golang.org/x/sys/unix](https://pkg.go.dev/golang.org/x/sys/unix) will be used because the `syscall` package of the standard library is deprecated.
+
+The limits to be used are:
+
+- `RLIMIT_AS`: virtual memory
+- `RLIMIT_CPU`: CPU time
+- `RLIMIT_DATA`: data and heap size
+- `RLIMIT_FSIZE`: size of the files created by the process
+- `RLIMIT_NOFILE`: number of files opened by the process
+- `RLIMIT_STACK`: stack size
+
+For a more detailed explanation of each limit see here: https://linux.die.net/man/2/prlimit.
+
+If time allows it (and is deemed necessary) an alternative implementation will be attempted, which is to `fork()` the server process, set the resource limits in place using the `setrlimit` system call and then `exec()` to the given job, avoiding the race condition described above. This would be straightforward to do in other languages such as C, but Go makes it more complicated as it doesn't expose `fork()` directly because it interferes with its goroutine scheduling system if not done within some constraints.
```

Comment:
This is a requirement, because as you mentioned, unless you handle re-exec, it's not safe.